clang/lib/Sema/SemaTemplateInstantiateDecl.cpp

```diff
@@ -6174,7 +6174,12 @@ static void processSYCLKernel(Sema &S, FunctionDecl *FD, MangleContext &MC) {
   if (S.LangOpts.SYCLIsDevice) {
     S.ConstructOpenCLKernel(FD, MC);
   } else if (S.LangOpts.SYCLIsHost) {
-    CXXRecordDecl *CRD = (*FD->param_begin())->getType()->getAsCXXRecordDecl();
+    QualType KernelParamTy = (*FD->param_begin())->getType();
+    const CXXRecordDecl *CRD;
+    if (KernelParamTy->isReferenceType())
+      CRD = KernelParamTy->getPointeeCXXRecordDecl();
+    else
+      CRD = KernelParamTy->getAsCXXRecordDecl();
```

Comment:
This part fixes the assertion but does not generate the subgroup attribute on host for AST and IR. So all tests associated with this PR will fail.  

I am debugging it and trying to understand why this does not work.

if (S.LangOpts.SYCLIsHost) {// work on Host 
    S.ConstructOpenCLKernel(FD, MC);

Reply:
It might be because CRD is ptr-to-const.   You may need a const_cast in the first assignment.

Reply:
I tried this:
if (KernelParamTy->isReferenceType())
-      CRD = KernelParamTy->getPointeeCXXRecordDecl();
+      CRD = const_cast<CXXRecordDecl *>(KernelParamTy->getPointeeCXXRecordDecl());